test/quantized_ops/test_quantized_matmul.py

```diff
@@ -97,9 +106,90 @@ def test_q_linear_hlo(self):
 
       output = torch.ops.xla.quantized_matmul(x, w_int, scaler)
       hlo = torch_xla._XLAC._get_xla_tensors_hlo([output])
-      print(hlo)
       self.assertTrue(re.search(r'bf16.*dot.*bf16.*s8', hlo) is not None)
 
+  def test_int4_pack_unpack(self):
+    x = torch.randint(-8, 7, (2, 12), dtype=torch.int8)
+    orig_dtypes = [
+        torch.int8,
+    ]
+    for dtype in orig_dtypes:
+      data = x.to(dtype)  # make sure the tensor is within the range of dtype.
+      packed = pack_4bit(data, dtype=dtype)
+      unpacked = unpack_4bit(packed, dtype=dtype)
+      self.assertTrue(packed.dtype == dtype)
+      self.assertTrue(unpacked.dtype == dtype)
+      self.assertTrue(torch.equal(data, unpacked))
+
+  def test_int4_per_channel_matmul(self):
+    weight = torch.randint(-8, 7, (4, 2)).to(torch.int8)
+    weight_scaler = torch.randn(4).to(torch.bfloat16)
+    x = torch.ones(3, 2).to(torch.bfloat16)
+
+    torch_out = torch.ops.xla.quantized_matmul(x, weight, weight_scaler)
+
+    x = x.to(device)
+    weight = weight.to(device)
+    weight_scaler = weight_scaler.to(device)
+
+    xla_out = torch.ops.xla.quantized_matmul(
+        x, weight, weight_scaler, int4_weight=True)
+
+    hlo = torch_xla._XLAC._get_xla_tensors_hlo([xla_out])
+    self.assertTrue(re.search(r'bf16.*dot.*bf16.*s4', hlo) is not None)
+
+    # Dot with int4 weight is only supported on TPU
+    if xr.device_type() == 'TPU':
+      self.assertGreater(
+          self._calc_cosine_dist(xla_out.cpu(), torch_out), 0.999999)
+
+  def test_int4_per_channel_linear_module(self):
+    m = M(5, 8)
+    x = torch.randn(3, 5)
+    out_fp = m(x)
+    m.replace_with_xla_quantized_matmul(n_bit=4)
+    out_quant = m(x)
+    self.assertGreater(self._calc_cosine_dist(out_fp, out_quant), 0.99)
+
+    # Dot with int4 weight is only supported on TPU
+    if xr.device_type() == 'TPU':
+      m = m.to(device)
+      x = x.to(device)
+      out_quant_xla = m(x)
+      self.assertGreater(
+          self._calc_cosine_dist(out_quant_xla.cpu(), out_quant), 0.999999)
+
+  @unittest.skip(
+      "Don't run, just an example for blockwise quant with int4 weight.")
+  def test_int4_blockwise_matmul(self):
```

Comment:
you should add it under `example` dir instead of in a test file.

Reply:
ok, let me remove it now.